Source/DafnyCore/Resolver/PreType/PreTypeResolve.cs

```diff
@@ -218,6 +218,63 @@ public enum Type2PreTypeOption { GoodForInference, GoodForPrinting, GoodForBoth
       return decl;
     }
 
+    /// <summary>
+    /// Returns the non-newtype ancestor of "cecl".
+    /// </summary>
+    public static TopLevelDecl AncestorDecl(TopLevelDecl decl) {
+      while (decl is NewtypeDecl newtypeDecl) {
+        var parent = newtypeDecl.BasePreType.Normalize();
```

Comment:
This part looks very similar to what is happening around line 251 in `NewTypeAncestor` below. Why do we have to account for cycles in the latter, but not here?

Reply:
`NewTypeAncestor` below detects cycles as it goes along. This is needed somewhere in the resolver. In contrast, `AncestorDecl` assumes that there are no cycles, so it should only be called after cycle detection. I will add comments to this effect.